### PR TITLE
Involve CheckStopReadingFn class adding a chance to release resources gracefully

### PR DESCRIPTION
**Pull request by gabrywu**
closing #29072

### CODE REVIEW

**Comment by gabrywu** [comment]
hi, @Abacn can you help to review this PR?

**Review by kennknowles — APPROVED** [review]
This generally LGTM. I looked at the bug and it makes sense. I don't really have context on what the intended uses of "checkStopReadingFn" are anyhow.

**Review by Abacn — COMMENTED** [review]
Thanks. The only caveat I could imagine is that this would break existing users using `SerializableFunction<TopicPartition, Boolean>` to `setCheckStopReadingFn`. It would be nice that setCheckStopReadingFn could still accept ordinary SerializableFunction

also aware that there is no guarantee that tearDown will be called after a DoFn is done according to DoFnLifeCycle

**Comment by gabrywu** [comment]
> This generally LGTM. I looked at the bug and it makes sense. I don't really have context on what the intended uses of "checkStopReadingFn" are anyhow.

Beside official doc tells intended uses, we use it to simulate kafka consumer group.

**Comment by gabrywu** [comment]
> SerializableFunction<TopicPartition, Boolean>

OK, I will create a wrapper of SerializableFunction<TopicPartition, Boolean> to avoid breaking user experience.

**Comment by gabrywu** [comment]
@Abacn please help to review it

File: sdks/java/io/kafka/src/test/java/org/apache/beam/sdk/io/kafka/ReadFromKafkaDoFnTest.java [hunk 1]
```diff
@@ -432,7 +432,7 @@ public void testProcessElementWhenTopicPartitionIsStopped() throws Exception {
             makeReadSourceDescriptor(consumer)
                 .toBuilder()
                 .setCheckStopReadingFn(
-                    new SerializableFunction<TopicPartition, Boolean>() {
+                    new CheckStopReadingFn() {
```
Comment (Abacn) [inline sdks/java/io/kafka/src/test/java/org/apache/beam/sdk/io/kafka/ReadFromKafkaDoFnTest.java:435]: sorry for being late. Let's revert the change in test to make sure the compatibility is tested

**Review by Abacn — APPROVED** [review]
LGTM other than a minor comment

File: sdks/java/io/kafka/src/test/java/org/apache/beam/sdk/io/kafka/ReadFromKafkaDoFnTest.java (hunk 1 above)
Comment (Abacn) [inline sdks/java/io/kafka/src/test/java/org/apache/beam/sdk/io/kafka/ReadFromKafkaDoFnTest.java:435]:
```suggestion
                    new SerializableFunction<TopicPartition, Boolean>() {
```

**Comment by Abacn** [comment]
Test passed locally

Test | Duration | Result
-- | -- | --
testProcessElementWhenTopicPartitionIsStopped | 0.140s | passed

**Comment by gabrywu** [comment]
> Thanks. The only caveat I could imagine is that this would break existing users using `SerializableFunction<TopicPartition, Boolean>` to `setCheckStopReadingFn`. It would be nice that setCheckStopReadingFn could still accept ordinary SerializableFunction
> 
> also aware that there is no guarantee that tearDown will be called after a DoFn is done according to DoFnLifeCycle

can you give more details of this case? I find that parts of codes within tearDown run actually. Why?

**Comment by Abacn** [comment]
Generally it will run. In the case there is an Exception in DoFn or runner side, for example, or pipeline get cancelled, teardown may not run. This was mostly about do not rely on teardown for data integrity management. Clean up resources in teardown are fine

**Comment by gabrywu** [comment]
Yes, per my understanding, teardown is used to clean up resources. Because dip don't have corresponding close method like flink, which is used to cleanup buffered data.
